### ISSUE TITLE
SQLalchemy: storing ParameterData object with "nan" value of an attribute raises an exception

### ISSUE THREAD

**Issue by yakutovicha** [aiidateam/aiida_core#1498]
To reproduce the problem one can do the following:

```
In [1]: from aiida.orm.data.parameter import ParameterData
In [2]: n = ParameterData(dict={"a":float("nan")})
In [3]: n.store()
```
The error message looks as follows:
```
DataError                                 Traceback (most recent call last)
<ipython-input-10-5a3ff09580fc> in <module>()
----> 1 p.store()

/project/.local/lib/python2.7/site-packages/aiida/orm/implementation/general/node.pyc in store(self, with_transaction)
   1550
   1551             # call implementation-dependent store method
-> 1552             self._db_store(with_transaction)
   1553
   1554             # Set up autogrouping used by verdi run

/project/.local/lib/python2.7/site-packages/aiida/orm/implementation/sqlalchemy/node.pyc in _db_store(self, with_transaction)
    641                 try:
    642                     # aiida.backends.sqlalchemy.get_scoped_session().commit()
--> 643                     session.commit()
    644                 except SQLAlchemyError as e:
    645                     #print "Cannot store the node. Original exception: {" \

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/session.pyc in commit(self)
    894                 raise sa_exc.InvalidRequestError("No transaction is begun.")
    895
--> 896         self.transaction.commit()
    897
    898     def prepare(self):

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/session.pyc in commit(self)
    453         self._assert_active(prepared_ok=True)
    454         if self._state is not PREPARED:
--> 455             self._prepare_impl()
    456
    457         if self._parent is None or self.nested:

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/session.pyc in _prepare_impl(self)
    433                 if self.session._is_clean():
    434                     break
--> 435                 self.session.flush()
    436             else:
    437                 raise exc.FlushError(

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/session.pyc in flush(self, objects)
   2110         try:
   2111             self._flushing = True
-> 2112             self._flush(objects)
   2113         finally:
   2114             self._flushing = False

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/session.pyc in _flush(self, objects)
   2228         except:
   2229             with util.safe_reraise():
-> 2230                 transaction.rollback(_capture_exception=True)
   2231
   2232     def bulk_save_objects(

/project/.local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.pyc in __exit__(self, type_, value, traceback)
     58             exc_type, exc_value, exc_tb = self._exc_info
     59             self._exc_info = None   # remove potential circular references
---> 60             compat.reraise(exc_type, exc_value, exc_tb)
     61         else:
     62             if not compat.py3k and self._exc_info and self._exc_info[1]:

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/session.pyc in _flush(self, objects)
   2192             self._warn_on_events = True
   2193             try:
-> 2194                 flush_context.execute()
   2195             finally:
   2196                 self._warn_on_events = False

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.pyc in execute(self)
    371                     self.dependencies,
    372                     postsort_actions):
--> 373                 rec.execute(self)
    374
    375     def finalize_flush_changes(self):

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.pyc in execute(self, uow)
    530                              uow.states_for_mapper_hierarchy(
    531                                  self.mapper, False, False),
--> 532                              uow
    533                              )
    534

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.pyc in save_obj(base_mapper, states, uowtransaction, single)
    176         _emit_insert_statements(base_mapper, uowtransaction,
    177                                 cached_connections,
--> 178                                 mapper, table, insert)
    179
    180     _finalize_insert_update_commands(

/project/.local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.pyc in _emit_insert_statements(base_mapper, uowtransaction, cached_connections, mapper, table, insert, bookkeeping)
    802                 else:
    803                     result = cached_connections[connection].\
--> 804                         execute(statement, params)
    805
    806                 primary_key = result.context.inserted_primary_key

/project/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.pyc in execute(self, object, *multiparams, **params)
    912                 type(object))
    913         else:
--> 914             return meth(self, multiparams, params)
    915
    916     def _execute_function(self, func, multiparams, params):

/project/.local/lib/python2.7/site-packages/sqlalchemy/sql/elements.pyc in _execute_on_connection(self, connection, multiparams, params)
    321
    322     def _execute_on_connection(self, connection, multiparams, params):
--> 323         return connection._execute_clauseelement(self, multiparams, params)
    324
    325     def unique_params(self, *optionaldict, **kwargs):

/project/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.pyc in _execute_clauseelement(self, elem, multiparams, params)
   1008             compiled_sql,
   1009             distilled_params,
-> 1010             compiled_sql, distilled_params
   1011         )
   1012         if self._has_events or self.engine._has_events:

/project/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.pyc in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1144                 parameters,
   1145                 cursor,
-> 1146                 context)
   1147
   1148         if self._has_events or self.engine._has_events:

/project/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.pyc in _handle_dbapi_exception(self, e, statement, parameters, cursor,context)
   1339                 util.raise_from_cause(
   1340                     sqlalchemy_exception,
-> 1341                     exc_info
   1342                 )
   1343             else:

/project/.local/lib/python2.7/site-packages/sqlalchemy/util/compat.pyc in raise_from_cause(exception, exc_info)
    201     exc_type, exc_value, exc_tb = exc_info
    202     cause = exc_value if exc_value is not exception else None
--> 203     reraise(type(exception), exception, tb=exc_tb, cause=cause)
    204
    205 if py3k:

/project/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.pyc in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1137                         statement,
   1138                         parameters,
-> 1139                         context)
   1140         except Exception as e:
   1141             self._handle_dbapi_exception(

/project/.local/lib/python2.7/site-packages/sqlalchemy/engine/default.pyc in do_execute(self, cursor, statement, parameters, context)
    448
    449     def do_execute(self, cursor, statement, parameters, context=None):
--> 450         cursor.execute(statement, parameters)
    451
    452     def do_execute_no_params(self, cursor, statement, context=None):

DataError: (psycopg2.DataError) invalid input syntax for type json
LINE 1: ...08T14:09:25.725847+00:00'::timestamptz, 1, false, '{"a": NaN...
                                                             ^
DETAIL:  Token "NaN" is invalid.
CONTEXT:  JSON data, line 1: {"a": NaN...
 [SQL: 'INSERT INTO db_dbnode (uuid, type, label, description, ctime, mtime, nodeversion, public, attributes, extras, dbcomputer_id, user_id) VALUES (%(uuid)s, %(type)s, %(label)s, %(description)s, %(ctime)s, %(mtime)s, %(nodeversion)s, %(public)s, %(attributes)s, %(extras)s, %(dbcomputer_id)s, %(user_id)s) RETURNING db_dbnode.id'] [parameters: {'user_id': 2, 'ctime': datetime.datetime(2018, 5, 8, 14, 9, 25, 725830, tzinfo=<UTC>), 'uuid': u'7700b986-5a09-422c-be21-8b9fe67ac97a', 'label': '', 'extras': '{}', 'mtime': datetime.datetime(2018, 5, 8,14, 9, 25, 725847, tzinfo=<UTC>), 'attributes': '{"a": NaN}', 'dbcomputer_id': None, 'nodeversion': 1, 'type': 'data.parameter.ParameterData.', 'public': False, 'description': ''}]
```

The same code works fine for the Django backend.

**Comment by giovannipizzi** [comment]
On Django, when you get it back, do you get NaN or something else? (I would expect the latter)
Also, can NaN be encoded into JSON? I don't think so, so in this case we can't do much (except documenting that NaN are not accepted, and maybe making a check before storing (even if this slows down storage so maybe just documentation?)

**Comment by giovannipizzi** [comment]
NaN are not supported by the PostgreSQL backend. See https://www.postgresql.org/docs/9.5/datatype-json.html, table 8-23.
This will soon be the case also for Django once we move to JSONB also there.
I think this should become a 'documentation' issue + some checking.
Tasks:
- [ ] please document that `NaN` and `infinity` are not allowed.
- [ ] improve `clean_value` to complain if `NaN` or `infinity` are found

**Comment by sphuber** [comment]
Fixed in PR #2295
